core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1000,6 +1016,46 @@ class ReplicaManager(val config: KafkaConfig,
     result
   }
 
+  case class SomeReplicaView(isLeader: Boolean,
+                             endpoint: Node,
+                             logOffset: Long,
+                             lastCaughtUpTimeMs: Long) extends ReplicaView
+
+  case class SomePartitionView(replicas: util.Set[ReplicaView],
+                               leader: util.Optional[ReplicaView]) extends PartitionView
+
+  def findPreferredReadReplica(tp: TopicPartition, clientMetadata: ClientMetadata, replicaId: Int): Option[Int] = {
+    val partition = getPartitionOrException(tp, expectLeader = false)
+
+    if (partition.leaderReplicaIfLocal.isDefined) {
+      if (Request.isValidBrokerId(replicaId)) {
+        // Don't look up preferred for follower fetches via normal replication
+        Option.empty
+      } else if (clientMetadata.equals(ClientMetadata.NO_METADATA)) {
+        // Return the leader if no metadata given
+        partition.leaderReplicaIdOpt
+      } else {
+        val replicaEndpoints = metadataCache.getPartitionReplicaEndpoints(tp.topic(), tp.partition(), new ListenerName(clientMetadata.listenerName))
+        val replicaInfoSet: Set[ReplicaView] = partition.allReplicas.map(
+          replica => SomeReplicaView(
+            isLeader = partition.leaderReplicaIdOpt.exists(leaderId => leaderId.equals(replica.brokerId)),
```

Comment:
@hachikuji is this a reasonable way to test if the local replica is the leader?

Reply:
I changed it to a default field based on the set of ReplicaView-s. If we really don't want it as a member of the interface I can move it to a static. It is convenient to quickly get the leader optional from the PartitionView

Reply:
Moved to a static method on ReplicaSelector